some preliminary work for a copy-delete type subscription

hufnagel opened the pull request:
Normally there are only two subscription modes supported, copy or move.
We want to implement a delayed move, consisting of a copy first, followed by
a delete from the source site once the workflow creating the data is done.

These are some prelimary changes for this. The main changes are in the schema,
adding a delete_blocks column to dbsbuffer_dataset_subscription and a
deleted column to dbsbuffer_block. To make this work, we also needed to
add a FK constrained dataset_id column to dbsbuffer_block, which then
forced us to make some code changes to properly fill this.

These code changes then triggered some neccessary cleanup work. Tried to
keep it as minimal as possible though for now, more is to come.

hufnagel committed:
some preliminary work for a copy-delete type subscription

Normally there are only two subscription modes supported, copy or move.
We want to implement a delayed move, consisting of a copy first, followed by
a delete from the source site once the workflow creating the data is done.

These are some prelimary changes for this. The main changes are in the schema,
adding a delete_blocks column to dbsbuffer_dataset_subscription and a
deleted column to dbsbuffer_block. To make this work, we also needed to
add a FK constrained dataset_id column to dbsbuffer_block, which then
forced us to make some code changes to properly fill this.

These code changes then triggered some neccessary cleanup work. Tried to
keep it as minimal as possible though for now, more is to come.

Also fix a few unit tests to work with the changes.

hufnagel commented:
@ticoann , can you review ? I don't want to necessarily have this merged yet, but I would like for someone to have a look to see if this makes sense. This is actually functional, it did survive a Tier0 replay with many datasets and blocks inserted into DBS3 and PhEDEx. No idea about breaking unit tests though...

hufnagel commented:
Btw, this also will break the WMCore/src/python/WMCore/Operations/MigrateFileBlocks.py  script as already discussed as it removes one of the DAOs called by that script.

hufnagel commented:
Another btw, if it helps get this merged quicker I could also split this into smaller pull requests. Some of the cleanup work was triggered by trying to make the other changes work, but is actually independent of these changes. Would be more work though, so if there is a chance to get this whole thing tested/merged relatively quickly I'd rather do that.

hufnagel commented:
Just updated the patch a bit, removed a few spurious import statements that got left from testing.

ticoann commented on src/python/WMComponent/DBS3Buffer/MySQL/NewSubscription.py:
```diff
@@ -32,13 +32,24 @@ def execute(self, datasetID, subscriptionInfo,
+                isMove = 1
+        else:
+                isMove = 0
+                if subscriptionInfo.get('DeleteFromSource', False):
```
This (DeleteFromSource) is set when spec is generated, right?

hufnagel replied:
Yes, it's part of the subscription information that is set in the spec. I can use it as-is from the Tier0 as I construct this myself in code. If we want to use it for normal requests some more code will be needed. But for now this just adds the capability, if not set (which is the default) it won't do anything.

ticoann commented:
Hi Dirk, the code looked fine to me. 
Aside from code clean up and schema changes,
So SetBlockStaus replaced by SetBlockClosed.
It seems that original SetBlockStatus dose multiple things but actual usages seems to be just updating the status. (When this gets called always the block exist, correct?)

One other thing is it might be better to update unit test as well. 
There are 3 packages running this in unittest. 
DBS3Buffer_t, DBSBuffer_t, and PhEDExInjector_t.
I think I would just remove DBSBuffer_t unittest all together but the other too should be updated. (It seems one of the current unittest is broken though - PhEDExInjectorSubscriber_t)

Thanks,

hufnagel commented:
Yes, the only time SetBlockStatus was called was for updating a block status to Closed and at that time the block existed. The only other place in the code it's called is the old DBSUpload component polling cycle code (which you said isn't used anymore) and in the MigrateFileBlocks.py ops script (where I am still waiting for an answer if anyone is still using it).

hufnagel commented:
I can look at the unit tests, removing the DBSBuffer_t ones and making sure the others work.

hufnagel commented:
@ticoann , worked on the unit tests. They are a mess (were before as well), but I removed the DBSBuffer_t directory and all unit tests in DBS3Buffer_t and PhEDExInjector_t work now (well, the DBSUpload_t one does not work, but it's a generic error message I get on all their unit tests, nothing related to my changes, the one error related to my changes I removed).

ticoann commented:
@amaltaro, Alan, you said this patch works fine, right?

amaltaro commented:
Not really :) I said I need to redeploy an agent with this patch (since it touches DB schema).
I'm finishing ReqMgr validation in Seangchan's private machine and then I'll test this one.
Give me a couple of days please.

amaltaro commented:
It's good to go! I ran 8 different flavors of workflows and everything is ok with the agent components, phedex injection and dbs registration.
One question though that I noticed only now, should we run these test workflows with automatic phedex subscriptions? If so, then I need to run a few more tests...

amaltaro commented:
Just in case, a few wfs with automatic phedex subscription did well.
